operator/src/main/java/org/keycloak/operator/controllers/KeycloakServiceMonitorDependentResource.java

```diff
@@ -0,0 +1,71 @@
+package org.keycloak.operator.controllers;
+
+import static org.keycloak.operator.controllers.KeycloakDeploymentDependentResource.managementEndpoint;
+import static org.keycloak.operator.crds.v2alpha1.CRDUtils.LEGACY_MANAGEMENT_ENABLED;
+import static org.keycloak.operator.crds.v2alpha1.CRDUtils.METRICS_ENABLED;
+import static org.keycloak.operator.crds.v2alpha1.CRDUtils.configuredOptions;
+
+import org.keycloak.operator.Constants;
+import org.keycloak.operator.Utils;
+import org.keycloak.operator.crds.v2alpha1.deployment.Keycloak;
+import org.keycloak.operator.crds.v2alpha1.deployment.spec.ServiceMonitorSpec;
+
+import io.fabric8.kubernetes.api.model.networking.v1.Ingress;
+import io.fabric8.openshift.api.model.monitoring.v1.ServiceMonitor;
+import io.fabric8.openshift.api.model.monitoring.v1.ServiceMonitorBuilder;
+import io.javaoperatorsdk.operator.api.config.informer.Informer;
+import io.javaoperatorsdk.operator.api.reconciler.Context;
+import io.javaoperatorsdk.operator.api.reconciler.dependent.DependentResource;
+import io.javaoperatorsdk.operator.processing.dependent.kubernetes.CRUDKubernetesDependentResource;
+import io.javaoperatorsdk.operator.processing.dependent.kubernetes.KubernetesDependent;
+import io.javaoperatorsdk.operator.processing.dependent.workflow.Condition;
+
+@KubernetesDependent(
+      informer = @Informer(labelSelector = Constants.DEFAULT_LABELS_AS_STRING)
+)
+public class KeycloakServiceMonitorDependentResource extends CRUDKubernetesDependentResource<ServiceMonitor, Keycloak> {
+
+    public static class ReconcilePrecondition implements Condition<Ingress, Keycloak> {
+        @Override
+        public boolean isMet(DependentResource<Ingress, Keycloak> dependentResource, Keycloak primary,
+                             Context<Keycloak> context) {
+            var opts = configuredOptions(primary);
+            if (Boolean.parseBoolean(opts.get(LEGACY_MANAGEMENT_ENABLED)) || !Boolean.parseBoolean(opts.getOrDefault(METRICS_ENABLED, "false")))
+                return false;
+
+            return ServiceMonitorSpec.get(primary).isEnabled();
+        }
+    }
+
+    @Override
+    protected ServiceMonitor desired(Keycloak primary, Context<Keycloak> context) {
+        var endpoint = managementEndpoint(primary, context, true);
+        var meta = primary.getMetadata();
+        var spec = ServiceMonitorSpec.get(primary);
+        return new ServiceMonitorBuilder()
+              .withNewMetadata()
+                .withName(meta.getName())
+                .withNamespace(meta.getNamespace())
```

Comment:
Do you know if users will want custom labels and annotations on these? If so it could be added as a follow-on task under #15395

Reply:
Possibly yes, we have had a request for this in the Infinispan Operator in the past: https://github.com/infinispan/infinispan-operator/issues/1777.

I have created the sub-issue #42626 :+1: